src/build.d

```diff
@@ -895,7 +895,15 @@ auto detectModel()
     if (detectOS == "solaris")
         uname = ["isainfo", "-n"].execute.output;
     else if (detectOS == "windows")
-        uname = ["wmic", "OS", "get", "OSArchitecture"].execute.output;
+    {
+        version (Windows)
+        {
+            import core.sys.windows.winbase;
```

Comment:
I would add:
```d
version (D_LP64)
    return "64";
else
{
    ...
```
so that we don't even call `IsWow64Process` on 64 bits, we already know we're 64-bit.

Reply:
Or even `version (Win64) return "64";`?

Reply:
This has to be a runtime check so that you can build a 64-bit executable using a 32-bit host compiler(?)

Reply:
Ah yes. I suppose you can use a 32-bit compiler on a 64-bit system.

Reply:
Shouldn't a 32-bit dmd detect that it's on a 64-bit Windows machine and set the version appropriately?

Reply:
```D
import std.stdio;

void main()
{
    version (Win32)
        writeln("win32");
    version (Win64)
        writeln("win64");
    version (D_LP64)
        writeln("D_LP64");
}
```

On my 64-bit windows machine, running a 32-bit dmd compiler, I get this output:
```
win32
```

If I add `-m64`, then I get this:
```
win64
D_LP64
```

So I guess the question is, do we want the default to match the model that the developer used to compile `build.d`, or do we want the default to always use 64 on 64-bit machines and 32 otherwise?  In the end, the default doesn't matter to much, it's just the default.

Reply:
If we were to use `version` here for defaults, why not extend this to other platforms as well and just use this:
```d
auto detectModel()
{
   version (D_LP64) return "64";
   else return "32";
}
```

Reply:
This is becoming a bit of a Chinese whispers situation.

- It is kind of weird how a compile-time condition (OS `version`) is translated into a runtime string value, which is then used to make decisions. I'm not sure if this is an artifact from a literal translation from makefiles, or some kind of future-proofing for cross-compilation.
- My understanding that for Windows we want to target the same bitness as the OS, not compiled executable, so this needs to be a runtime check.
- The old method, calling `wmic.exe`, was needlessly expensive (running a separate program) and fragile (broke Digger, and output depends on the current locale). We should be using Windows APIs instead.
- Because D and Windows come in only 32-bit and 64-bit flavors, and because you can't run 64-bit executables on 32-bit Windows, I suggested checking `D_LP64` as an optimization: if the current executable is 64-bit, then we can infer that the OS **must** also be 64-bit, so go ahead and target 64-bit without checking the OS bitness.

Reply:
@CyberShadow I see now, a clever optimization.  If we were to use that, I would probably suggest a comment like this:
```D
// If build.d is compiled as 64-bit, then we MUST be on a 64-bit system.  The inverse is not guaranteed.
version (D_LP64)
    return "64";
else
{
```